guacamole/src/main/java/org/apache/guacamole/tunnel/InterceptedStream.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.guacamole.protocol.GuacamoleStatus;
+
+/**
+ * A simple pairing of the index of an intercepted Guacamole stream with the
+ * stream-type object which will produce or consume the data sent over the
+ * intercepted Guacamole stream.
+ *
+ * @author Michael Jumper
+ * @param <T>
+ *     The type of object which will produce or consume the data sent over the
+ *     intercepted Guacamole stream. Usually, this will be either InputStream
+ *     or OutputStream.
+ */
+public class InterceptedStream<T extends Closeable> {
+
+    /**
+     * The index of the Guacamole stream being intercepted.
+     */
+    private final String index;
+
+    /**
+     * The stream which will produce or consume the data sent over the
+     * intercepted Guacamole stream.
+     */
+    private final T stream;
+
+    /**
+     * The exception which prevented the stream from completing successfully,
+     * if any. If the stream completed successfully, or has not encountered any
+     * exception yet, this will be null.
+     */
+    private GuacamoleException streamError = null;
+
+    /**
+     * Creates a new InterceptedStream which associated the given Guacamole
+     * stream index with the given stream object.
+     *
+     * @param index
+     *     The index of the Guacamole stream being intercepted.
+     *
+     * @param stream
+     *     The stream which will produce or consume the data sent over the
+     *     intercepted Guacamole stream.
+     */
+    public InterceptedStream(String index, T stream) {
+        this.index = index;
+        this.stream = stream;
+    }
+
+    /**
+     * Returns the index of the Guacamole stream being intercepted.
+     *
+     * @return
+     *     The index of the Guacamole stream being intercepted.
+     */
+    public String getIndex() {
+        return index;
+    }
+
+    /**
+     * Returns the stream which will produce or consume the data sent over the
+     * intercepted Guacamole stream.
+     *
+     * @return
+     *     The stream which will produce or consume the data sent over the
+     *     intercepted Guacamole stream.
+     */
+    public T getStream() {
+        return stream;
+    }
+
+    /**
+     * Reports that this InterceptedStream did not complete successfully due to
+     * the given GuacamoleException, which could not be thrown at the time due
+     * to asynchronous handling of the stream contents.
+     *
+     * @param streamError
+     *     The exception which prevented the stream from completing
+     *     successfully.
+     */
+    public void setStreamError(GuacamoleException streamError) {
+        this.streamError = streamError;
+    }
+
+    /**
+     * Reports that this InterceptedStream did not complete successfully due to
+     * an error described by the given status code and human-readable message.
+     * The error reported by this call can later be retrieved as a
+     * GuacamoleStreamException by calling getStreamError().
```

Comment:
It's just a `GuacamoleException`, but the subclass of `GuacamoleException` that will be set due to an invocation of this function will be `GuacamoleStreamException`.